### PR TITLE
Remove _log_softmax

### LINKED ISSUE #76433: Remove _log_softmax/_softmax in favor of log_softmax and softmax respectively.
### 🚀 The feature, motivation and pitch

Similar rationale to https://github.com/pytorch/pytorch/issues/73135

The only difference between `_softmax` and `softmax` is that `softmax` takes an optional dtype argument, while `_softmax` takes a `half_to_float` argument. However, this `half_to_float` arg is not needed in the backwards pass, and even if it were, we could simply reconstruct it from the argument dtypes.

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L427

So, to remove these, we can simply turn the current `_softmax` and `_log_softmax` into C++ functions (not operators), and mostly preserve the same logic as today. The backwards pass would need to be modified slightly to take account of the cast.

cc: @albanD  @zou3519 @ngimel @ezy

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Chillee**
Fixes https://github.com/pytorch/pytorch/issues/76433

Will also remove `_softmax`, and rename `_log_softmax_backward_data` and `_softmax_backward_data` in follow-up commits.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

**Comment by ngimel** [comment]
Nice cleanup! Can you check with static runtime if they still need their softmax implementation?

**Comment by mikeiovine** [comment]
Removing this from static runtime is fine, but you'll also need to remove the [generated unit test](https://github.com/pytorch/pytorch/blob/master/benchmarks/static_runtime/test_generated_ops.cc#L112)

File: test/test_ops.py [hunk 1]
```diff
@@ -512,6 +512,7 @@ def _compare_out(transform, *, compare_strides_and_data_ptrs=True):
                 final_strides = _extract_strides(out)
                 final_ptrs = _extract_data_ptrs(out)
 
+                print(expected, out)
```
Comment (ezyang) [inline test/test_ops.py:515]: remove me

**Review by ezyang — APPROVED** [review]

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -1874,9 +1874,9 @@
 - name: log_sigmoid_forward(Tensor self) -> (Tensor output, Tensor buffer)
   self: log_sigmoid_backward(grad, self, buffer)
 
-- name: _log_softmax(Tensor self, int dim, bool half_to_float) -> Tensor
+- name: log_softmax.int(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
   self: _log_softmax_backward_data(grad, result, dim, self.scalar_type())
```
Comment (ngimel) [inline tools/autograd/derivatives.yaml:1878]: The fact that tests don't break means that autograd engine implicitly casts grad to the self type? Nice I guess...

**Review by ngimel — APPROVED** [review]

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (Chillee) [reply tools/autograd/derivatives.yaml:1878]: It's actually that `_log_softmax_backward_data` takes in "input_dtype" as an argument. So, if `self` is fp32 and `dtype` is fp64, then we have

```
_log_softmax_backward_data(grad: fp64, result: fp64, dim, self.scalar_type(): fp32)
```
So the `_log_softmax_backward_data` kernel casts from whatever grad dtype it's given to the input dtype.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -2833,24 +2833,23 @@
...
 - func: _log_softmax.out(Tensor self, int dim, bool half_to_float, *, Tensor(a!) out) -> Tensor(a!)
+  variants: function
+
+- func: log_softmax.int_out(Tensor self, int dim, ScalarType? dtype=None, *, Tensor(a!) out) -> Tensor(a!)
   structured: True
   dispatch:
     CPU: log_softmax_cpu_out
     CUDA: log_softmax_cuda_out
 
+- func: log_softmax.Dimname(Tensor self, Dimname dim, *, ScalarType? dtype=None) -> Tensor
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:2850]: nit the diff is a bit hard to read here since things were moved around. This one is not changing right?

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:1878]: It does:
https://github.com/pytorch/pytorch/blob/f6bbecf8b503eb6562d52b66124a49b45a57f14a/torch/csrc/autograd/engine.cpp#L704-L706

**Review by albanD — APPROVED** [review]
SGTM

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (Chillee) [reply aten/src/ATen/native/native_functions.yaml:2850]: Correct.
